graph_weather/models/atmorep/model/multiformer.py

```diff
@@ -0,0 +1,247 @@
+            mask_dict (dict, optional): A dictionary where each key corresponds to a field
+                                         and the value is the mask tensor of shape [B, T, N]. Default is None.
+
+        Returns:
+            dict: A dictionary of ensemble predictions for each field after applying
+                  both cross-attention and spatiotemporal attention.
+        """
+        # Get basic MultiFormer output
+        encoded_features, multi_scale_features = {}, {}
+
+        for field_name, field_data in x_dict.items():
+            field_mask = mask_dict.get(field_name) if mask_dict else None
+            encoded, features = self.field_transformers[field_name](field_data, field_mask)
+            encoded_features[field_name] = encoded
+            multi_scale_features[field_name] = features
+
+        # Apply cross-attention between fields
+        cross_attended_features = {field: feat.clone() for field, feat in encoded_features.items()}
+
+        for field1 in self.config.input_fields:
+            for field2 in self.config.input_fields:
+                if field1 != field2:
+                    cross_key = f"{field2}_to_{field1}"
+                    cross_attn = self.cross_attentions[cross_key]
+                    cross_attended_features[field1] += cross_attn(
+                        encoded_features[field1], encoded_features[field2]
+                    )
+
+        # Apply spatiotemporal attention
+        spatiotemporal_features = {}
+        for field_name, field_feature in cross_attended_features.items():
+            spatiotemporal_features[field_name] = self.spatiotemporal_attns[field_name](
+                field_feature
+            )
+
+        # Decode each field
+        decoded_features = {}
+        for field_name, field_encoded in spatiotemporal_features.items():
+            decoded = self.decoders[field_name](field_encoded, multi_scale_features[field_name])
+            decoded_features[field_name] = decoded
+
+        # Apply ensemble prediction heads
+        predictions = {}
+        for field_name, field_decoded in decoded_features.items():
+            field_preds = []
+            for head in self.ensemble_heads[field_name]:
+                # Reshape to the original spatial dimensions
+                B, T, N, D = field_decoded.shape
+                H = W = int(math.sqrt(N) * self.config.patch_size)
+                pred = head(field_decoded)
+                pred = rearrange(
+                    pred, "b t (h w) c -> b t (h c) w", h=int(math.sqrt(N)), w=int(math.sqrt(N))
+                )
+                pred = F.interpolate(pred, size=(H, W), mode="bilinear")
+                pred = rearrange(pred, "b t c w -> b t w c")
+                field_preds.append(pred)
+
+            # Stack ensemble predictions
+            predictions[field_name] = torch.stack(field_preds, dim=0)  # [E, B, T, H, W]
+
+        return predictions
+
+
+class DataParallelAtmoRep(nn.Module):
```

Comment:
This should be more left to the training script, not here. I would remove this class.